### DISCUSSION TITLE
How develop and test this package?

### DISCUSSION THREAD

**Discussion by campbead** [ThinkR-open/golem#1113] [Q&A]
I am working on modifying some of the functions in this package along with writing associate tests using `testthat`.  However, when I make a change to a function and write new tests, the unmodified versision of the function is being loaded.  I believe this is because of the `library(golem)` command in [helper-config.R](https://github.com/ThinkR-open/golem/blob/master/tests/testthat/helper-config.R).  However, if I disable the `library(golem)` command, the workflows fails before my tests execute.  

The only workaround I can see is renaming the functions I am changing to something not in the namespace of `golem`.  However, that seems less than ideal.  

I am wondering if someone could guide me towards a more appropriate method for developing and testing this package?  Please let me know if you need any additional information.

Thank you.

**Comment by ColinFay** [comment]
Hey, 

Yeah testing golem is a bit weird — the test suite moves to different folders and loads namespaces because well, golem is a package to manage package :) Testthat calls pkgload::load_all(), something that golem does too, and golem tries to unload namespaces so it can be a bit messy from time to time.

I usually recommend using a test that runs outside the current R session, for example via a new terminal or if you're an RStudio button with the 'test' button. Another thing you can try is restarting R between each test run, so that you're sure you're in a clean state. 

That being said, I do feel like the library(golem) should be removed from the helper config file.

**Comment by campbead** [comment]
I have been able to make progress the testing problems that I had been experiencing.  I'll describe my method since it varies from @ColinFay 's method above.  

I am using RStudio to create and edit functions and tests.  

- What didn't work for me:
    - Using the "Run Tests" button on the File Editor toolbar.
    - Running the "Test" button in the Build tab.  
- What did work for me:
   - In the RStudio console running the following:
```
pkgload::load_all()
testthat::test_file("tests/testthat/test-file.R")
```

Any further changes to the function would require reloading the functions with `pkgload::load_all()`. Additionally, I can use this process to test the entire package with `devtools::test()`. However, it's worth noting that this method can be a bit clunky because RStudio prompt you to switch projects during the testing process.

I'm curious to know if there are other methods for testing and would appreciate any ideas for improving the testing process for this package. 

Thank you for your assistance.
